pkg/controller/networkpolicyrecommendation/util.go

```diff
@@ -0,0 +1,326 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to get recommendation result with id %s: %v", id, err)
+	}
+	resultTimeCreated, err := time.Parse(clickHouseTimeFormat, resultTimeCreatedStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse timeCreated of the result for %s: %v", id, err)
+	}
+	recommendedNetworkPolicy := &crdv1alpha1.RecommendedNetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "rnp-",
+		},
+		Spec: crdv1alpha1.RecommendedNetworkPolicySpec{
+			Id:          id,
+			Type:        resultType,
+			TimeCreated: metav1.NewTime(resultTimeCreated),
+			Yamls:       resultYamls,
+		},
+	}
+	return recommendedNetworkPolicy, nil
+}
+
+func getClickHouseSecret(client kubernetes.Interface, namespace string) (username []byte, password []byte, err error) {
+	secret, err := client.CoreV1().Secrets(namespace).Get(context.TODO(), "clickhouse-secret", metav1.GetOptions{})
+	if err != nil {
+		return username, password, fmt.Errorf("error when finding the ClickHouse secret. Error: %v", err)
+	}
+	username, ok := secret.Data["username"]
+	if !ok {
+		return username, password, fmt.Errorf("error when getting the ClickHouse username")
+	}
+	password, ok = secret.Data["password"]
+	if !ok {
+		return username, password, fmt.Errorf("error when getting the ClickHouse password")
+	}
+	return username, password, nil
+}
+
+func connectClickHouse(url string) (*sql.DB, error) {
+	var connect *sql.DB
+	// Open the database and ping it
+	var err error
+	connect, err = openSql("clickhouse", url)
+	if err != nil {
+		return connect, fmt.Errorf("failed to open ClickHouse: %v", err)
+	}
+	if err := connect.Ping(); err != nil {
+		if exception, ok := err.(*clickhouse.Exception); ok {
+			return connect, fmt.Errorf("failed to ping ClickHouse: %v", exception.Message)
+		} else {
+			return connect, fmt.Errorf("failed to ping ClickHouse: %v", err)
+		}
+	}
+	return connect, nil
+}
+
+func setupClickHouseConnection(client kubernetes.Interface, namespace string) (connect *sql.DB, err error) {
+	serviceIP, servicePort, err := getServiceAddr(client, "clickhouse-clickhouse", namespace, "tcp")
+	if err != nil {
+		return connect, fmt.Errorf("error when getting the ClickHouse Service address: %v", err)
+	}
+	endpoint := fmt.Sprintf("tcp://%s:%d", serviceIP, servicePort)
+
+	// Connect to ClickHouse and execute query
+	username, password, err := getClickHouseSecret(client, namespace)
+	if err != nil {
+		return nil, err
+	}
+	url := fmt.Sprintf("%s?debug=false&username=%s&password=%s", endpoint, username, password)
+	connect, err = connectClickHouse(url)
+	if err != nil {
+		return nil, fmt.Errorf("error when connecting to ClickHouse, %v", err)
+	}
+	return connect, nil
+}
+
+func getPolicyRecommendationResult(client kubernetes.Interface, id string, namespace string) (recommendedNetworkPolicy *crdv1alpha1.RecommendedNetworkPolicy, err error) {
+	connect, err := setupClickHouseConnection(client, namespace)
```

Comment:
can we cache / resue connection within the controller instead of reestablishing connection each time db check is needed?

Reply:
Updated, thanks!